README.md

# IoT Spotify Clock

My goal with this project was to become better acquainted with an [RGB LED Matrix](https://www.adafruit.com/product/5036) from AdaFruit. I wanted to be able to show two different things: 

* The time
* What I'm currently listening to on Spotify

I originally wanted to be able to swap between many different displays (e.g., weather), but simplified to better understand the capabilities and constraints before taking that step. 

The (current) final output is a voice-activated display. It starts by showing the time while waiting to update from Spotify. When a song is played, it displays 
the title and artist. At any point, I can swap between the clock display and the Spotify display. I can also ask to change the color of the primary display. 

![image](https://media.giphy.com/media/h0UTMkgSwvAoSCCqaU/giphy.gif?cid=790b76111a91e0a7643c755ff39486fa9b73d34ee3ace350&rid=giphy.gif)

The swapping of display and color is done using an Alexa skill: 


<img src="https://github.com/ahope/iot_clock/blob/master/alexa.png" width="200">

# Requirements

### Hardware

* Adafruit RGB LED Matrix: https://www.adafruit.com/product/2278
* Adafruit MatrixPortal board:  https://www.adafruit.com/product/4745

This project was based on AdaBox 016-- a 64x32 RGB LED Matrix, with the MatrixPortal M4 board. The board is CircuitPython friendly, but can also be used with the Arduino IDE for lower-level programming. 

### Software/Third Parties

* adafruit.io (MQTT Broker)
* IFTTT 
* Alexa (Either a device, or the app)
* Alexa Skill 
* Spotify
* Abode (not required-- I just added this for fun!)

## System Overview

Here's an overview of the entire system: 

<img src="https://github.com/ahope/iot_clock/blob/master/iot_clock_diagram_1.png" width="250">

To change the display via Alexa: 

* Tell Alexa to either change the color or the display. 
* The Alexa backend sends a message to the adafruit.io MQTT webhook. 
* The display is subscribed to the MQTT broker, and when notified of an update, either changes the display or the color of the display. 

To update the message/song on the display: 

* There are 2 applets on IFTTT: 
   * Spotify: Every hour, notify adafruit.io what songs have been played. 
   * Abode: When the system status changes, notify adafruit.io of the update.


## Details

#### Basic setup: 

* Initialize the display
* Initialize the network 
* Initialize the MQTT client
* Initialize the colors and fonts
* Set up the MQTT subscriptions for color and display changing

#### Enabling Clock Display: `set_display_clock()`

* Create a display Group
* Show the Group
* Create a label for the display
* Get the time (via the network-- ensure the chip clock matches the correct time zone)
* Figure out the correct placement based on size
* Add the label to the Group
* Remember that the clock is being displayed

#### Enabling Spotify Display: `set_display_spotify()`

* Create a display Group
* Create artist/title labels
* Set the default text for the two labels
* Set the correct placement for the labels
* Show the display group
* Remember that the spotify stuff is being displayed

#### The `run` loop: 

* If the clock is showing: 
    - Check if the time has changed (next minute-- roughly check every minute)
* If the spotify feed is showing: 
    - Scroll the labels
* Check for new MQTT messages
    - Check that we're still connected
    - Check if any messages have come in
* Sleep for a few seconds and keep repeating

#### When a message is received: 

* If it's a color update: 
    - Save the new current color. 
    - Next time the clock or spotify display updates, it changes the labels to the new color. 
* If it's a display update: 
    - Call either `set_display_clock` or `set_display_spotify`
* If it's a Spotify update: 
    - Parse out the title and the artist
    - Update the artist and title labels with the new info

# The Future

As I always say, I build to think. Playing around with technology helps me to think about the possibilities. Some ideas 
that have come up: 

* Well, of course, making the code nicer and designed to more easily add new displays. 
* Add a "Timer" feature
   *  In the short term, a clock count-down
   *  In the longer term, an hour-glass timer.
   *  In the LONG longer term, the clock vs hour-glass depends on the orientation of the display
*  Add in different layouts for different screen orientations
*  Create a fun "screen-saver" display that is really just playing with pixels.
*  Probably a different project, but create a Euchre scorekeeper! 
   *  I'm actually really excited about this one. Definitely want to incorporate this with Alexa for voice control. 
   *  Keep track of current score, current trump, maybe even display card values for the current trump. 


    
# Helpful Links: 

* Adafruit MatrixPortal Library https://circuitpython.readthedocs.io/projects/matrixportal/en/latest/api.html#adafruit-matrixportal-matrixportal
* Adafruit IO MQTT API https://io.adafruit.com/api/docs/mqtt.html#adafruit-io-mqtt-api
* Adafruit MiniMQTT Library https://circuitpython.readthedocs.io/projects/adafruit-circuitpython-minimqtt/en/latest/api.html
* Matrix Portal Scoreboard project https://github.com/seantibor/matrix_portal_scoreboard
* Adafruit Matrix Portal library overview and such https://learn.adafruit.com/adafruit-matrixportal-m4?view=all#circuitpython-ble
* CircuitPython libraries for Matrix Portal M4 https://circuitpython.org/board/matrixportal_m4/


history.md
# Timeline/Notes of how the project evolved


### First pass

The clock was up and running fairly quickly. The hard part was figuring out how to connect to Spotify. After adding in some sample code to test the network connection and understand the request library, I realized that we can't do user authorization for Spotify because there is no UI/redirect URL. I *was* going consider utilizing MQTT as a second phase/upgrade, but I won't be able to get the access token via the ESP. 

### Second pass (PLAN)

Set up the matrix to utilize MQTT. The easy peasy first here: get the matrix subscribing to the MQTT broker. Set up IFTTT to listen to my Spotify listening, and publish to the broker. Then, the display updates. 

Done! Also set up a channel to change the color of the display, and to display when the security system is set to "Home" status. 

### Third pass

Now I've got a display with a clock and some messages providing info about Spotify and the alarm system. 

I want to add in the weather display. Adafruit has a cool weather display. I started by copying and pasting the relevant code into the same code file. 

#### Challenge 1: Different Projects using Different Classes/Constructs

The Clock and Scoreboard implementations use MatrixPortal for graphics, while hte Weather used the lower level Matrix class for graphics. Took me a little while to figure out the relationship between all of them. I understood the theoretical relationship from the diagram on the library overview page, but I had to dig into the source code to really understand. But, once I did that it was obvious and easy. 

Thus, I got the weather part of the app running too. 

#### Challenge 2: Memory!!! 

Finally! I shoved enough on the board and was sloppy enough with the planning that we ran out of memory! I knew this would happen, but wasn't quite sure when. So, now that we've got all the pieces, time to organize it so we can swap things out rather than keep in memory all the time. 

### Finalizing

I'm having tons of ideas for now, and just want something working with loose ends tied up. So I'm not going to debug the complex weather thing, I'm just going to make what I have work by cleaning it up. 

That means I end up with: 

* 2 Displays: One for Spotify, one for the clock
* 3 data sources: 
    - Display: which thing to display
    - color: which color to show things in
    - spotify: the messages coming from MQTT/IFTTT




# Debugging

I started thinking about how I go about debugging this project. There are a few options: 

* Software Profiling
* Hardware Profiling via a GPIO pin and an oscilloscop
* SWD-- ARM's version of JTAG. 

## Software Profiling

This is the embedded version of ```printf```. Write a ```profile``` function that stores some info about what is happening when into an array. For every function you want to profile, call the ```profile``` function when it gets called. When the array is full, stop running and dump out the data to be analyzed.

* This is a great way to see how long it takes a function to run, as well as how many times it's called. 
* It could be really fun to pull this up in R or a Jupyter notebook and visualize the data over time. Not sure it's worth it though! 
* This isn't really non-intrusive profiling. It takes space, and so won't be helpful to help debug memory issues. 
* This approach is easy to turn profiling on/off without having to touch many lines of code. 

## Hardware Profiling

In this case, we use one of the GPIO pins on the board. For every function we want to profile, turn the GPIO pin on at the start of the function, and off when the function is over. Then, we read the voltage on that IO pin, and pull the stats from that. 

* We can use a different IO pin for different functions, and that could help isolate the profile channels. 
* We need a way to read the data on that pin. The easiest is probably with an oscilloscope. Then, the features of your oscilloscope limit what you can do with the data. Some oscilloscopes can read multiple channels at a time; some oscilloscopes can record traces. I know my oscilloscope is at school, so while I love the idea of this approach, it's not always the most available. 

## SWD: Serial Wire Debug

This is a feature available on ARM chips. It provides a serial interface to the board for debugging. It's like JTAG, but instead of requiring 4 wires, it only requires 2, so it's less complex on the board. 

I've got a bit to figure out here, but one upside is that there *is* a Python library [https://github.com/pfalcon/PySWD] that likely helps. 



code.py
# Scoreboard matrix display
# uses AdafruitIO to set scores and team names for a scoreboard
# Perfect for cornhole, ping pong, and other games

import time
import board
import terminalio
import displayio
from adafruit_matrixportal.matrixportal import MatrixPortal
from adafruit_display_text.label import Label
from adafruit_matrixportal.network import Network
from adafruit_bitmap_font import bitmap_font
import adafruit_minimqtt.adafruit_minimqtt as MQTT
import adafruit_esp32spi.adafruit_esp32spi_socket as socket
from secrets import secrets
# import openweather_graphics 

# --- Display setup ---
matrixportal = MatrixPortal(status_neopixel=board.NEOPIXEL, debug=False)
network = matrixportal.network
network.connect()

display = matrixportal.graphics.display

mqtt = MQTT.MQTT(
    broker=secrets.get("mqtt_broker"),
    username=secrets["aio_username"],
    password=secrets["aio_key"],
    port=1883,
)


DISPLAY_CLOCK = 1
DISPLAY_SPOTIFY = 2


CURRENT_DISPLAY = DISPLAY_SPOTIFY## DISPLAY_CLOCK

RED_COLOR = 0xAA0000
TURQUOISE_COLOR = 0x00FFAA
PINK_COLOR = 0xFF0088

text_colors = [TURQUOISE_COLOR, PINK_COLOR, RED_COLOR]
current_text_color = 0

MQTT.set_socket(socket, network._wifi.esp)


last_data = {}

group = None# = displayio.Group()  # Create a Group
bitmap = None # = displayio.Bitmap(64, 32, 2)  # Create a bitmap object,width, height, bit depth
color = None #= displayio.Palette(4)  # Create a color palette

# Create a TileGrid using the Bitmap and Palette
tile_grid = None #= displayio.TileGrid(bitmap, pixel_shader=color)

font = bitmap_font.load_font("/IBMPlexMono-Medium-24_jep.bdf")
clock_label = Label(font)
artist_label = None
title_label = None

small_font = bitmap_font.load_font("/fonts/Arial-12.bdf")
medium_font = bitmap_font.load_font("/fonts/Arial-14.bdf")

BLINK = True




def set_display_clock(): 
    # global matrixportal
    global clock_label
    global group
    global bitmap
    global color
    global display
    # matrixportal.add_text(text_color=0xFF8800,
    #                       text_position=(30,5))
    # now = time.localtime() 
    # matrixportal.set_text(now[3], CLOCK_DISPLAY_TEXT_INDEX)

    # --- Drawing setup ---
    group = displayio.Group()  # Create a Group
    bitmap = displayio.Bitmap(64, 32, 2)  # Create a bitmap object,width, height, bit depth
    color = displayio.Palette(4)  # Create a color palette
    color[0] = 0x000000  # black background
    color[1] = 0xFF0000  # red
    color[2] = 0xCC4000  # amber
    color[3] = 0x85FF00  # greenish

    # Create a TileGrid using the Bitmap and Palette
    tile_grid = displayio.TileGrid(bitmap, pixel_shader=color)
    group.append(tile_grid)  # Add the TileGrid to the Group
    display.show(group)

    clock_label = Label(font)
    clock_label.text = "Yo!"
    clock_label.x = 3
    clock_label.y = 3   

    update_time()
    group.append(clock_label)
    CURRENT_DISPLAY = DISPLAY_CLOCK


def set_display_spotify():
    global title_label
    global artist_label
    group = displayio.Group()
    artist_label = Label(small_font)
    title_label = Label(medium_font)

    artist_label.text = "Artist"
    title_label.text = "Song Title Goes Here"

    bbx, bby, bbwidth, bbh = title_label.bounding_box
    # Center the label
    print(title_label.bounding_box)
    title_label.x = 1 #3round(display.width / 2 - bbwidth / 2)
    title_label.y = -1*bby ##display.height // 2

    # bbx, bby, bbwidth, bbh = artist_label.bounding_box
    artist_label.x = 1
    artist_label.y = title_label.y + bbh + 1 ##32 - bbh

    group.append(title_label)
    group.append(artist_label)

    display.show(group)
    CURRENT_DISPLAY = DISPLAY_SPOTIFY


    # matrixportal.add_text(text_wrap=10, 
    #                 text_maxlen=25, 
    #                 text_position=(2, 15),
    #                 scrolling=False)
    # matrixportal.set_text("waiting for update", 0)

    pass

def spotify_update(message):
    global title_label
    global artist_label
    index = message.find("|")
    if (index >= 0):
        title_text = message[0:index]
        artist_text = message[index + 1:]
        title_label.text = title_text
        artist_label.text = artist_text
        title_label.color = text_colors[current_text_color]

def change_display(message):
    if (message == "spotify"):
        set_display_spotify()
    if (message == "clock"):
        set_display_clock()
    pass

def message_received(client, topic, message):
    print("Received {} for {}".format(message, topic))
    if (topic == "ahslaughter/feeds/matrix-display-feeds.color"): 
        color_update(message)
    if (topic == "ahslaughter/feeds/matrix-display-feeds.spotify"):
        spotify_update(message)
    if (topic == "ahslaughter/feeds/matrix-display-feeds.which-display"):
        change_display(message)
    else: 
        matrixportal.set_text(message, 0)
        matrixportal.scroll_text()


localtime_refresh = None
# weather_refresh = None


def update_time(*, hours=None, minutes=None, show_colon=False):
    global clock_label
    global display
    global text_colors
    global current_text_color

    print("updating the time")
    now = time.localtime()  # Get the time values we need
    # print(now)
    # matrixportal.set_text('{0}:{1}'.format(now[3]%12, now[4]), 
    #     CLOCK_DISPLAY_TEXT_INDEX)

    if hours is None:
        hours = now[3]
    if hours >= 18 or hours < 6:  # evening hours to morning
        clock_label.color = color[1]
    else:
        clock_label.color = color[3]  # daylight hours
    if hours > 12:  # Handle times later than 12:59
        hours -= 12
    elif not hours:  # Handle times between 0:00 and 0:59
        hours = 12

    if minutes is None:
        minutes = now[4]

    if BLINK:
        colon = ":" if show_colon or now[5] % 2 else " "
    else:
        colon = ":"

    clock_label.text = "{hours}{colon}{minutes:02d}".format(
        hours=hours, minutes=minutes, colon=colon
    )

    clock_label.color = text_colors[current_text_color]


    bbx, bby, bbwidth, bbh = clock_label.bounding_box
    # Center the label
    clock_label.x = round(display.width / 2 - bbwidth / 2)
    clock_label.y = display.height // 2

    # print(clock_label.bounding_box)
    # print(display.width)

    # if DEBUG:
    print("Label bounding box: {},{},{},{}".format(bbx, bby, bbwidth, bbh))
    print("Label x: {} y: {}".format(clock_label.x, clock_label.y))
    print("label: {}".format(clock_label.text))


# def update_spotify():



def get_last_data(feed):
    feed_url = feeds.get(feed)
    return last_data.get(feed_url)


def subscribe():
    try:
        mqtt.is_connected()
    except MQTT.MMQTTException:
        mqtt.connect()
    mqtt.on_message = message_received

def color_update(which):
    global current_text_color
    print("Changing the color")
    current_text_color = int(which)
    # matrixportal.set_text_color(text_colors[int(which)], 0)



subscribe()
mqtt.subscribe("ahslaughter/feeds/matrix-display-feeds.spotify") ##"ahslaughter/feeds/spotify")
mqtt.subscribe("ahslaughter/feeds/matrix-display-feeds.color") ##"ahslaughter/feeds/spotify")
mqtt.subscribe("ahslaughter/feeds/matrix-display-feeds.which-display")
mqtt.on_message = message_received

last_check = None
localtime_refresh = None


def update_clock():
    global last_check
    if last_check is None or time.monotonic() > last_check + 3600:
        try:
            update_time(
                show_colon=True
            )  # Make sure a colon is displayed while updating
            network.get_local_time()  # Synchronize Board's clock to Internet
            last_check = time.monotonic()
        except RuntimeError as e:
            print("Some error occured, retrying! -", e)

    update_time()



def update_mqtt_messages(): 
    try:
        mqtt.is_connected()
        mqtt.loop()
    except (MQTT.MMQTTException, RuntimeError):
        network.connect()
        mqtt.reconnect()

# set_display_clock()
set_display_spotify()

while True:
    if (CURRENT_DISPLAY == DISPLAY_CLOCK):
        update_clock()
        
    update_mqtt_messages()    

    
    # if (CURRENT_DISPLAY == DISPLAY_WEATHER):
    #     update_weather()

    time.sleep(3)

metro_matrix_clock.py
# Metro Matrix Clock
# Runs on Airlift Metro M4 with 64x32 RGB Matrix display & shield

import time
import board
import displayio
import terminalio
from adafruit_display_text.label import Label
from adafruit_bitmap_font import bitmap_font
from adafruit_matrixportal.network import Network
from adafruit_matrixportal.matrix import Matrix

## Spotify Authorization
url = "https://accounts.spotify.com/api/token"
headers = {}
data = {}

# Encode as Base64
message = f"{clientId}:{clientSecret}"
messageBytes = message.encode('ascii')
base64Bytes = base64.b64encode(messageBytes)
base64Message = base64Bytes.decode('ascii')


headers['Authorization'] = f"Basic {base64Message}"
data['grant_type'] = "client_credentials"

r = requests.post(url, headers=headers, data=data)
token = r.json()['access_token']

# Step 2 - Use Access Token to call playlist endpoint

playlistId = "myPlaylistId"
playlistUrl = f"https://api.spotify.com/v1/playlists/{playlistId}"
headers = {
    "Authorization": "Bearer " + token
}

res = requests.get(url=playlistUrl, headers=headers)

print(json.dumps(res.json(), indent=2))

# bear = secrets['bearer_token']

# DATA_SOURCE = ('https://api.twitter.com/2/tweets/search/recent?query=#partyparrot to:blitzcitydiy')
# DATA_LOCATION = ["meta", "newest_id"]

# matrixportal = MatrixPortal(
#     url=DATA_SOURCE,
#     json_path=DATA_LOCATION,
#     status_neopixel=board.NEOPIXEL
# )




#### =============  CLOCK =========

BLINK = True
DEBUG = False

# Get wifi details and more from a secrets.py file
try:
    from secrets import secrets
except ImportError:
    print("WiFi secrets are kept in secrets.py, please add them there!")
    raise
print("    Metro Minimal Clock")
print("Time will be set for {}".format(secrets["timezone"]))

# --- Display setup ---
matrix = Matrix()
display = matrix.display
network = Network(status_neopixel=board.NEOPIXEL, debug=False)

# --- Drawing setup ---
group = displayio.Group()  # Create a Group
bitmap = displayio.Bitmap(64, 32, 2)  # Create a bitmap object,width, height, bit depth
color = displayio.Palette(4)  # Create a color palette
color[0] = 0x000000  # black background
color[1] = 0xFF0000  # red
color[2] = 0xCC4000  # amber
color[3] = 0x85FF00  # greenish

# Create a TileGrid using the Bitmap and Palette
tile_grid = displayio.TileGrid(bitmap, pixel_shader=color)
group.append(tile_grid)  # Add the TileGrid to the Group
display.show(group)

if not DEBUG:
    font = bitmap_font.load_font("/IBMPlexMono-Medium-24_jep.bdf")
else:
    font = terminalio.FONT

clock_label = Label(font)


def update_time(*, hours=None, minutes=None, show_colon=False):
    now = time.localtime()  # Get the time values we need
    if hours is None:
        hours = now[3]
    if hours >= 18 or hours < 6:  # evening hours to morning
        clock_label.color = color[1]
    else:
        clock_label.color = color[3]  # daylight hours
    if hours > 12:  # Handle times later than 12:59
        hours -= 12
    elif not hours:  # Handle times between 0:00 and 0:59
        hours = 12

    if minutes is None:
        minutes = now[4]

    if BLINK:
        colon = ":" if show_colon or now[5] % 2 else " "
    else:
        colon = ":"

    clock_label.text = "{hours}{colon}{minutes:02d}".format(
        hours=hours, minutes=minutes, colon=colon
    )
    bbx, bby, bbwidth, bbh = clock_label.bounding_box
    # Center the label
    clock_label.x = round(display.width / 2 - bbwidth / 2)
    clock_label.y = display.height // 2
    if DEBUG:
        print("Label bounding box: {},{},{},{}".format(bbx, bby, bbwidth, bbh))
        print("Label x: {} y: {}".format(clock_label.x, clock_label.y))


last_check = None
update_time(show_colon=True)  # Display whatever time is on the board
group.append(clock_label)  # add the clock label to the group

while True:
    if last_check is None or time.monotonic() > last_check + 3600:
        try:
            update_time(
                show_colon=True
            )  # Make sure a colon is displayed while updating
            network.get_local_time()  # Synchronize Board's clock to Internet
            last_check = time.monotonic()
        except RuntimeError as e:
            print("Some error occured, retrying! -", e)

    update_time()
    time.sleep(1)


spotify_clock.py


import time
import board
import terminalio
import displayio
from adafruit_matrixportal.matrixportal import MatrixPortal
from adafruit_display_text.label import Label
from adafruit_matrixportal.network import Network
from adafruit_bitmap_font import bitmap_font
import adafruit_minimqtt.adafruit_minimqtt as MQTT
import adafruit_esp32spi.adafruit_esp32spi_socket as socket
from secrets import secrets

# --- Display setup ---
matrixportal = MatrixPortal(status_neopixel=board.NEOPIXEL, debug=False)
display = matrixportal.graphics.display

# --- Network setup ---
network = matrixportal.network
network.connect()
network.get_local_time() 

# --- MQTT Setup ---
mqtt = MQTT.MQTT(
    broker=secrets.get("mqtt_broker"),
    username=secrets["aio_username"],
    password=secrets["aio_key"],
    port=1883,
)

MQTT.set_socket(socket, network._wifi.esp)

MQTT_URL_SPOTIFY = "ahslaughter/feeds/matrix-display-feeds.spotify"
MQTT_URL_COLOR = "ahslaughter/feeds/matrix-display-feeds.color"
MQTT_URL_WHICH_DISPLAY = "ahslaughter/feeds/matrix-display-feeds.which-display"

MQTT_TOPIC_COLOR = "color"
MQTT_TOPIC_WHICH_DISPLAY = "which-display"
MQTT_TOPIC_SPOTIFY = "spotify"

## --- Display Options ---
DISPLAY_CLOCK = 1
DISPLAY_SPOTIFY = 2

current_display = DISPLAY_SPOTIFY

scroll_delay = 0.03

NUM_COLORS = 5
RED = 0xAA0000
TURQUOISE = 0x00FFAA
PINK = 0xFF0088
YELLOW = 0xFFFF00
ORANGE = 0xFF7722

text_colors = [RED, ORANGE, YELLOW, TURQUOISE, PINK]
current_text_color = 0

## --- Fonts ---
font = bitmap_font.load_font("/IBMPlexMono-Medium-24_jep.bdf")
small_font = bitmap_font.load_font("/fonts/Arial-12.bdf")
medium_font = bitmap_font.load_font("/fonts/Arial-14.bdf")

## --- MQTT data and info ---
last_data = {}




## --- Clock data and info ---
group = None
bitmap = None
color = None

# Create a TileGrid using the Bitmap and Palette
tile_grid = None 

clock_label = Label(font)
artist_label = None
title_label = None

BLINK = True

last_time_check = None

TIME_STRING_FORMAT = "{hours}{colon}{minutes:02d}"
current_time_string = None


def set_display_clock(): 
    print("setting up the clock display")
    global clock_label
    global group
    global bitmap
    global color
    global display
    global current_display

    # --- Drawing setup ---
    group = displayio.Group()  # Create a Group
    bitmap = displayio.Bitmap(64, 32, 2)  # Create a bitmap object,width, height, bit depth

    # Create a TileGrid using the Bitmap and Palette
    tile_grid = displayio.TileGrid(bitmap, pixel_shader=color)
    group.append(tile_grid)  # Add the TileGrid to the Group
    display.show(group)

    clock_label = Label(font) 

    update_time()
    group.append(clock_label)
    current_display = DISPLAY_CLOCK
    print("curent display is hte clock")


def set_display_spotify():
    global title_label
    global artist_label
    global current_display
    group = displayio.Group()
    artist_label = Label(small_font)
    title_label = Label(medium_font)

    now = time.localtime()  # Get the time values we need
    print(now)
    artist_label.text = '{0}:{1}'.format(now[3], now[4])
    title_label.text = "Waiting for update..."

    bbx, bby, bbwidth, bbh = title_label.bounding_box
    # Center the label
    print(title_label.bounding_box)
    title_label.x = 1 
    title_label.y = -1*bby 

    artist_label.x = 1
    artist_label.y = title_label.y + bbh + 1 

    group.append(title_label)
    group.append(artist_label)

    display.show(group)

    mqtt.subscribe(MQTT_URL_SPOTIFY)
    mqtt.add_topic_callback(MQTT_TOPIC_SPOTIFY, spotify_update)

    current_display = DISPLAY_SPOTIFY

def clear_display_spotify():
    mqtt.unsubscribe(MQTT_URL_SPOTIFY)
    mqtt.remove_topic_callback(MQTT_TOPIC_SPOTIFY)


def spotify_update(message):
    global title_label
    global artist_label
    print("got a spotify update!")
    last_data.push(message)
    index = message.find("|")
    if (index >= 0):
        title_text = message[0:index]
        artist_text = message[index + 1:]
        title_label.text = title_text
        artist_label.text = artist_text
        title_label.color = text_colors[current_text_color]

def scroll_spotify():
    if len(last_data) == 0: 
        artist_label.text = current_time_string

    title_label.color = text_colors[current_text_color]

    if title_label.width > display.width: 
        # Run a loop until the label is offscreen again and leave function
        for _ in range(title_label.width):
            title_label.x = title_label.x - 1
            time.sleep(scroll_delay)
        title_label.x = display.width
        for _ in range(display.width):
            title_label.x = title_label.x - 1
            time.sleep(scroll_delay)

    if artist_label.width > display.width:
        # Run a loop until the label is offscreen again and leave function
        for _ in range(artist_label.width):
            artist_label.x = artist_label.x - 1
            time.sleep(scroll_delay)
        artist_label.x = display.width
        for _ in range(display.width):
            artist_label.x = artist_label.x - 1
            time.sleep(scroll_delay)

def change_display(message):
    if (message != "spotify"):
        clear_display_spotify()

    if (message == "spotify"):
        set_display_spotify()
        return
    if (message == "clock"):
        set_display_clock()
        return
    print(message)

def message_received(client, topic, message):
    print("Received {} for {}".format(message, topic))
    if (topic == "ahslaughter/feeds/matrix-display-feeds.color"): 
        color_update(message)
    if (topic == "ahslaughter/feeds/matrix-display-feeds.spotify"):
        spotify_update(message)
    if (topic == "ahslaughter/feeds/matrix-display-feeds.which-display"):
        change_display(message)
    else: 
        matrixportal.set_text(message, 0)
        matrixportal.scroll_text()


### Checks the current time
### Populates the global current_time_string with the current time. 
def update_time(*, hours=None, minutes=None, show_colon=False):
    global current_time_string

    now = time.localtime()  # Get the time values we need

    if hours is None:
        hours = now[3]
    if hours > 12:  # Handle times later than 12:59
        hours -= 12
    elif not hours:  # Handle times between 0:00 and 0:59
        hours = 12

    if minutes is None:
        minutes = now[4]

    if BLINK:
        colon = ":" if show_colon or now[5] % 2 else " "
    else:
        colon = ":"

    current_time_string = TIME_STRING_FORMAT.format(
        hours=hours, minutes=minutes, colon=colon
    )



def get_last_data(feed):
    feed_url = feeds.get(feed)
    return last_data.get(feed_url)


def subscribe():
    try:
        mqtt.is_connected()
    except MQTT.MMQTTException:
        mqtt.connect()
    
    mqtt.subscribe(MQTT_URL_COLOR)
    mqtt.subscribe(MQTT_URL_WHICH_DISPLAY)
    mqtt.on_message = message_received
    mqtt.add_topic_callback(MQTT_TOPIC_COLOR, color_update)
    mqtt.add_topic_callback(MQTT_TOPIC_WHICH_DISPLAY, change_display)

def color_update(which):
    global current_text_color
    print("color_update")
    new_color_index = int(which)
    if new_color_index < NUM_COLORS:
        current_text_color = new_color_index


def update_clock():
    clock_label.text = current_time_string

    clock_label.color = text_colors[current_text_color]
    bbx, bby, bbwidth, bbh = clock_label.bounding_box
    # Center the label
    clock_label.x = round(display.width / 2 - bbwidth / 2)
    clock_label.y = display.height // 2


def update_mqtt_messages(): 
    try:
        mqtt.is_connected()
        mqtt.loop()
    except (MQTT.MMQTTException, RuntimeError):
        network.connect()
        mqtt.reconnect()

subscribe()
# set_display_clock()
set_display_spotify()

while True:
    if last_time_check is None or time.monotonic() > last_time_check + 3600:
        network.get_local_time()  # Synchronize Board's clock to Internet
        last_time_check = time.monotonic()
    update_time(show_colon=True)

    if (current_display == DISPLAY_CLOCK):
        update_clock()
        
    if (current_display == DISPLAY_SPOTIFY):
        scroll_spotify()

    update_mqtt_messages()    

    time.sleep(1)